pyomo/gdp/plugins/bigm.py

```diff
@@ -255,117 +250,143 @@ def _apply_to_impl(self, instance, **kwds):
+        # deactivate so the writers don't scream
         obj.deactivate()
 
-    def _transformDisjunctionData(self, obj, transBlock, bigM, index):
+    def _transform_disjunctionData(self, obj, bigM, index, xorConstraint=None,
+                                   transBlock=None):
         if not obj.active:
-            return  # Do not process a deactivated disjunction
-        parent_component = obj.parent_component()
-        transBlock.disjContainers.add(parent_component)
-        orConstraint = self._getXorConstraint(parent_component)
+            return  # Do not process a deactivated disjunction 
+        # We won't have these arguments if this got called straight from
+        # targets. But else, we created them earlier, and have just been passing
+        # them through.
+        if transBlock is None:
+            # It's possible that we have already created a transformation block
+            # for another disjunctionData from this same container. If that's
+            # the case, let's use the same transformation block. (Else it will
+            # be really confusing that the XOR constraint goes to that old block
+            # but we create a new one here.)
+            if not obj.parent_component()._algebraic_constraint is None:
+                transBlock = obj.parent_component()._algebraic_constraint().\
+                             parent_block()
+            else:
+                transBlock = self._add_transformation_block(obj.parent_block())
+        if xorConstraint is None:
+            xorConstraint = self._add_xor_constraint(obj.parent_component(),
+                                                     transBlock)
 
         xor = obj.xor
         or_expr = 0
+        # Just because it's unlikely this is what someone meant to do...    
+        if len(obj.disjuncts) == 0:
+            raise GDP_Error("Disjunction %s is empty. This is " 
+                            "likely indicative of a modeling error."  %
+                            obj.getname(fully_qualified=True,
+                                        name_buffer=NAME_BUFFER))
         for disjunct in obj.disjuncts:
             or_expr += disjunct.indicator_var
             # make suffix list. (We don't need it until we are
             # transforming constraints, but it gets created at the
             # disjunct level, so more efficient to make it here and
-            # pass it down.
+            # pass it down.)
             suffix_list = self._get_bigm_suffix_list(disjunct)
+            arg_list = self._get_bigm_arg_list(bigM, disjunct)
             # relax the disjunct
-            self._bigM_relax_disjunct(disjunct, transBlock, bigM, suffix_list)
+            self._transform_disjunct(disjunct, transBlock, bigM, arg_list,
+                                     suffix_list)
+
         # add or (or xor) constraint
         if xor:
-            orConstraint.add(index, (or_expr, 1))
+            xorConstraint.add(index, expr=or_expr == 1)
```

Comment:
FYI: `.add(index, expr)` is legacy syntax.  The preferred syntax should be:
```python
xorConstraint[index] = or_expr == 1
```
or
```python
xorConstraint[index].set_value(or_expr == 1)
```